### ISSUE TITLE
Allow closures that can be called recursively

### ISSUE THREAD

**Issue by WorldSEnder** [rustwasm/wasm-bindgen#3172]
### Motivation

See https://github.com/yewstack/yew/issues/2989. There is an event handler that, as part of a user-defined function passed to the library, dispatches additional events to the same handler. This fails, because it re-enters the same event handler, which is not allowed.

### Proposed Solution

There should be an API that can be used to create safe, re-entrant, recursively callable functions that can be sent to js. On the surface level, there are two problems:
- `pub fn wrap(mut data: Box<T>)  -> Closure<T>` specifies that internally, the closure keeps a box. Since Boxes *own* their data, they contain a `Unique` and can not be safely exist twice on the same stack. One would have to change the internal management of the closure. Since it's working with `unsafe` code anyway, one could directly work with `NonNull<T>`.
- The API could only accept `Fn`, not `FnMut` or `FnOnce`, for similar reasons.

### Alternatives

I don't see any alternatives or workarounds that could be applied in user code, in general. For the specific case, a workaround is mentioned in the linked issue, but it would degrade performance and may surprise a lot of users.

**Comment by Liamolucko** [comment]
`Closure`s can already be called reentrantly, as long as they're `Closure<dyn Fn(...)>` rather than `Closure<dyn FnMut(...)>`.

Even though Yew already requires event listeners to implement `Fn`, it ends up creating `FnMut` closures anyway, because they're created via. `gloo_utils::EventListener::new_with_options` here:

https://github.com/yewstack/yew/blob/17f125c9ec8e7e526551bc1c6dd52b77f58a544a/packages/yew/src/dom_bundle/subtree_root.rs#L138-L153

That API takes an `FnMut`, and as such creates an `FnMut` closure. You'll have to use `add_event_listener` directly so that you can manually create an `Fn` closure and allow reentrancy.

**Comment by hamza1311** [comment]
I wonder if it'd be a good idea to make gloo's EventListener create an Fn closure, instead of FnMut one

**Comment by Liamolucko** [comment]
> I wonder if it'd be a good idea to make gloo's EventListener create an Fn closure, instead of FnMut one

Maybe. That'd make it impossible to create `EventListener`s from `FnMut`s, but I'm not sure how often you'd need to do that.

**Comment by WorldSEnder** [comment]
> `Closure`s can already be called reentrantly, as long as they're `Closure<dyn Fn(...)>` rather than `Closure<dyn FnMut(...)>`.

Ah. I had a quick look into the code and was under the impression that the same check was done for `Fn` that is done for `FnMut`.

https://github.com/rustwasm/wasm-bindgen/blob/6ed6883ae19d3e10d5e6715731ec886dd8830470/src/closure.rs#L573-L575

https://github.com/rustwasm/wasm-bindgen/blob/6ed6883ae19d3e10d5e6715731ec886dd8830470/src/closure.rs#L761-L763

I will take another look in the coming days to double check this and see if a quick patch in gloo would be sufficient.

**Comment by Liamolucko** [comment]
> Ah. I had a quick look into wasm_bindgen's code and was under the impression that the same check was done for `Fn` that is done for `FnMut`.
> 
> https://github.com/rustwasm/wasm-bindgen/blob/6ed6883ae19d3e10d5e6715731ec886dd8830470/src/closure.rs#L573-L575
> 
> https://github.com/rustwasm/wasm-bindgen/blob/6ed6883ae19d3e10d5e6715731ec886dd8830470/src/closure.rs#L761-L763

Right, the error message for `Fn` is pretty misleading there. Even though they both check for the pointer getting set to 0, on the JS side that happens under different circumstances. For `FnMut`, it gets set to 0 if the function gets called reentrantly *or* if the `Closure` is dropped, whereas for `Fn` it only gets set to 0 if the closure is dropped. So, for `Fn`, the error should really be 'closure invoked after being dropped'.

> But, as I said above, I don't think a reentrant call would be _sound_ with the current approach of re-creating a `Box<F>::from_raw`. After all, a `Box` _owns_ its data and _uniquely_ references its contents, so it must _not_ exist twice on the stack. Not sure if that actually matters in practice, but I'm almost sure miri would complain. See for example [this issue on rust unsafe code guidelines](https://github.com/rust-lang/unsafe-code-guidelines/issues/326)

I see what you mean, but `wasm_bindgen` doesn't use `Box::from_raw`, it just invokes the function through a raw pointer:

https://github.com/rustwasm/wasm-bindgen/blob/6ed6883ae19d3e10d5e6715731ec886dd8830470/src/closure.rs#L568-L588

Doing that while the `Box` is still around does seem dubious, but Miri seems fine with it in this snippet that does more-or-less the same thing:

```rust
fn main() {
    let mut x: Box<dyn Fn()> = Box::new(|| {});
    // use the intermediate `&mut` because that's how wasm_bindgen does it, just
    // in case that causes any problems.
    let ptr = &mut *x as *mut dyn Fn() as *const dyn Fn();
    unsafe {
        // alternate between using the box and the pointer to make sure there
        // are no problems with them both being used at once.
        x();
        (*ptr)();
        x();
        (*ptr)();
        x();
    }
}
```

**Comment by WorldSEnder** [comment]
Fixed downstream with yewstack/yew#3037, the discussed workaround works nicely. I've also seen that the error messages were updated, which is much appreciated! Going to close this here.

**Comment by hamza1311** [comment]
Worth noting that I also tried to solve this in gloo but the type check failed as `Closure::once` returns `FnMut` which can't be stored in the same type as `Fn`. The `EventListener` struct couldn't hold both types in one field
